File: 2164-two-best-non-overlapping-events/two-best-non-overlapping-events.py
class Solution(object):
    def maxTwoEvents(self, events):
        """
        :type events: List[List[int]]
        :rtype: int
        """

        # Sorting events based on starting times 
        events.sort()

        # Creating a list for suffix_maximum
        n = len(events)
        suffix_maximum = [0]*n
        suffix_maximum[n-1] = events[n-1][2]
        for i in range(n-2,-1,-1):
            suffix_maximum[i] = max(events[i][2],suffix_maximum[i+1])

        #Iterating intervals
        maximum_sum = suffix_maximum[0]
        for i in range(n-1):
            low = i+1
            high = n-1
            target = events[i][1]+1

            while high-low > 1:
                mid = (low + high)//2
                if events[mid][0] >= target:
                    high = mid
                else:
                    low = mid

            if events[low][0] >= target:
                maximum_sum = max(maximum_sum,events[i][2]+suffix_maximum[low])
            elif events[high][0] >= target:
                maximum_sum = max(maximum_sum,events[i][2] + suffix_maximum[high])

        return maximum_sum

        
        

File: 2164-two-best-non-overlapping-events/two-best-non-overlapping-events.cpp
class Solution {

public:
    int maxTwoEvents(vector<vector<int>>& events) {
        // Sorting events based on start times
        sort(events.begin(),events.end(),[](vector<int>&a,vector<int>&b){
            if(a[0] < b[0])return true;
            if(a[0] > b[0])return false;
            return a[2] > b[2];
        });

        // Creating a suffix array to store suffix maxes
        int n = events.size();
        vector<int>suffix(n);
        suffix[n-1] = events[n-1][2];
        for(int i = n-2 ; i >= 0 ; i--){
            suffix[i] = max(suffix[i+1],events[i][2]);
        } 

        // Binary search
        auto search_array = [](vector<vector<int>>&events,int target,int low,int high){
            while(high - low > 1){
                int mid = (low + high)>>1;
                if(events[mid][0] >= target){
                    high = mid;
                }else{
                    low = mid;
                }
            }

            if(events[low][0] >= target)return low;
            if(events[high][0] >= target)return high;
            return -1;
        };

        // Finding the max value of an interval that starts after each end time
        int max_answer = suffix[0];
        for(int i = 0 ; i < n-1 ; i++){
            int next_index = search_array(events,events[i][1]+1,i+1,n-1);
            if(next_index != -1){
                max_answer = max(max_answer,events[i][2] + suffix[next_index]);
            }
        }
        return max_answer;
    }
};